Add back missing methods in TcpClient, UdpClient, and TcpListener

ericeil opened the pull request:
Part of #11788 

Adds back the remaining methods of TcpClient, UdpClient, and TcpListener.

This PR does not include new tests for most of these.  I've done basic testing of the new TcpClient.Connect methods, and would like to get this merged so that @Priya91 can make use of those methods.  I'll file an issue to track adding tests for the rest of these.

@CIPop @stephentoub @davidsh

ericeil commented:
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please
@dotnet-bot Test Outerloop OSX Debug please

davidsh commented on src/System.Net.Sockets/ref/4.1.0/System.Net.Sockets.depproj:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
```
I thought we didn't need .depproj files anymore.

@ericstj You mentioned that we would do "package harvesting" to find the previous RTM version of a contract?

ericeil replied:
Yeah, I thought so too.  But without this, I get errors when building for netcore50, telling me I haven't implemented the new methods.  It seems like at least the API compat checker is *not* doing the "harvesting" thing, and instead is just using the new contract I just built.

weshaggard replied:
In general things should be harvested and we shouldn't need to add any depproj's. The only cases where we have needed to do it is for things like picking the older contracts for APICompat checks which sound like what you are hitting here. The other option we have is to just disable APICompat checks for the older builds, see https://github.com/dotnet/corefx/blob/71de1aac5cf3e5dc87714ae0cc829e1fa98f7c5d/src/System.IO.FileSystem/src/System.IO.FileSystem.csproj#L16.

davidsh commented on src/System.Net.Sockets/ref/4.1.0/project.json:
```diff
@@ -0,0 +1,12 @@
+      ]
+    }
+  }
+}
```
nit: missing newline at end of file.

ericeil replied:
fixed

davidsh commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.netstandard17.cs:
```diff
@@ -0,0 +1,187 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Why are we creating files with a ".netstandard17" suffix?

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Windows.netstandard17.cs:
```diff
@@ -0,0 +1,135 @@
+                                if (ipv6Socket != null)
+                                {
+                                    ipv6Socket.Close();
+                                }
```
Nit: could use C# 6 syntax here and elsewhere e.g.
```C#
ipv6Socket?.Close();
```

ericeil replied:
Good idea; done.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Windows.netstandard17.cs:
```diff
@@ -0,0 +1,135 @@
+                            break;
+                        }
+                    }
+
```
Nit: unnecessary blank line

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Windows.netstandard17.cs:
```diff
@@ -0,0 +1,135 @@
+
+                    catch (Exception ex)
+                    {
+                        if (ex is OutOfMemoryException)
```
Nit: could use C# 6 syntax for filters here, e.g.
```C#
catch (Exception ex) when (!(ex is OutOfMemoryException))
{
    lastex = ex;
}
```

ericeil replied:
Another good idea; done.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/UDPClient.netstandard17.cs:
```diff
@@ -0,0 +1,245 @@
+            //       returned from DNS. As a result, we defer the creation of the 
+            //       socket until the Connect method.
+            //
+            //CreateClientSocket();
```
Same question about commented out code

ericeil replied:
Same answer. :)

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Unix.cs:
```diff
@@ -290,7 +298,7 @@ private Task ConnectAsyncCore(IPAddress[] addresses, int port)
                         using (_disposing.Token.Register(o => ((Socket)o).Dispose(), s))
                         {
-                            await s.ConnectAsync(address, port).ConfigureAwait(false);
+                            await connect(s, address, port).ConfigureAwait(false);
```
Clever way to reuse the existing solution.

justinvp commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:
```diff
@@ -124,7 +124,39 @@ private IAsyncResult BeginConnect(IPAddress address, int port, AsyncCallback req
+        {
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Enter(NetEventSource.ComponentType.Socket, this, "BeginConnect", host);
```
Nit: `nameof(BeginConnect)`?

ericeil replied:
Fixed

justinvp commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:
```diff
@@ -124,7 +124,39 @@ private IAsyncResult BeginConnect(IPAddress address, int port, AsyncCallback req
+            IAsyncResult result = Client.BeginConnect(host, port, requestCallback, state);
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Exit(NetEventSource.ComponentType.Socket, this, "BeginConnect", null);
```
Nit: `nameof(BeginConnect)`?

justinvp commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:
```diff
@@ -124,7 +124,39 @@ private IAsyncResult BeginConnect(IPAddress address, int port, AsyncCallback req
+        {
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Enter(NetEventSource.ComponentType.Socket, this, "BeginConnect", addresses);
```
Nit: `nameof(BeginConnect)`?

justinvp commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:
```diff
@@ -124,7 +124,39 @@ private IAsyncResult BeginConnect(IPAddress address, int port, AsyncCallback req
+            IAsyncResult result = Client.BeginConnect(addresses, port, requestCallback, state);
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Exit(NetEventSource.ComponentType.Socket, this, "BeginConnect", null);
```
Nit: `nameof(BeginConnect)`?

justinvp commented on src/System.Net.Sockets/src/System/Net/Sockets/UDPClient.netstandard17.cs:
```diff
@@ -0,0 +1,245 @@
+        {
+            if (_cleanedUp)
+            {
+                throw new ObjectDisposedException(this.GetType().FullName);
```
Nit: `this.` isn't necessary. Applies throughout.

ericeil replied:
In this case, I like the explicit `this`.  For some reason a naked `GetType` call confuses me. :)

ericeil commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Windows.netstandard17.cs:
```diff
@@ -0,0 +1,135 @@
+                    }
+                    if (Socket.OSSupportsIPv6)
+                    {
+                        ipv6Socket = new Socket(AddressFamily.InterNetworkV6, SocketType.Stream, ProtocolType.Tcp);
```
Yes, though I'd like to address that separately, when I have better tests to validate the change. :)

stephentoub replied:
Ok.

ericeil commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Windows.netstandard17.cs:
```diff
@@ -0,0 +1,135 @@
+    {
+        private void ConnectCore(string hostname, int port)
+        {
+            // IPv6: We need to process each of the addresses return from
```
fixed

ericeil commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.netstandard17.cs:
```diff
@@ -0,0 +1,187 @@
+            //       performed DNS resolution and know which address
+            //       families we can use.
+            //
+            //InitializeClientSocket();
```
The comment was from the desktop code; I think it was meant to convey that this call is what we're *not* doing here.  I've removed it.

ericeil commented:
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please
@dotnet-bot Test Outerloop OSX Debug please

ericeil commented:
Added some commits to address feedback and fix the test failure.  Also added commit 14b355eddf8e7d6a3644b099caca7911845c6656, which fixes the System.Net.Sockets contract to use the newer Stream contract; I'd done that in the implementation, but apparently missed doing it in the contract, and @Priya91 needs this change as well for her work in progress.

ericeil commented:
@dotnet-bot test Innerloop CentOS7.1 Debug Build and Test please
[Previous failure](http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/centos7.1_debug_prtest/1268/) due to #12228.

ericeil commented:
@dotnet-bot test Innerloop Windows_NT Debug Build and Test please
[Build machine went offline](http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/windows_nt_debug_prtest/1207/)

ericeil commented:
@dotnet-bot test Innerloop Windows_NT Release Build and Test please
[Previous failure](http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/windows_nt_release_prtest/1203/) filed as #12427.

ericeil commented:
@dotnet-bot test OuterLoop Windows_NT Debug please
[Previous failure](http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/outerloop_windows_nt_debug_prtest/39/) was an odd build issue; while I research the possible cause, let's see if it happens consistently.

ericeil commented:
@weshaggard, do you have any idea what might be causing this build error?

> error MSB3030: Could not copy the file "D:\j\workspace\windows_nt_de---4526f5ff\bin\Windows_NT.AnyCPU.Debug\System.Net.Sockets\System.Net.Sockets.dll" because it was not found. [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Sockets\tests\FunctionalTests\System.Net.Sockets.Tests.csproj]

It seems to be intermittent; most of the builds succeeded.

weshaggard commented on src/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -61,9 +62,11 @@
     <Compile Include="System\Net\Sockets\SocketOptionName.cs" />
     <Compile Include="System\Net\Sockets\SocketShutdown.cs" />
     <Compile Include="System\Net\Sockets\TCPClient.cs" />
+    <Compile Condition="'$(TargetGroup)' != 'netcore50'" Include="System\Net\Sockets\TCPClient.netstandard17.cs" />
```
Can we group these into one ItemGroup with the condition on it instead of duplicating the conditions multiple places.

ericeil commented:
@dotnet-bot Test Outerloop Windows_NT Debug please
@dotnet-bot Test Outerloop Ubuntu14.04 Debug please
@dotnet-bot Test Outerloop OSX Debug please

ericeil commented:
The build error turned out to actually be due to an ApiCompat failure, because I was missing implementations for the methods I added to the contract at the last minute.  I thought I'd already added these, but just forgot to add them to the contract, and my own local build was successful because I didn't do a *re*build, which apparently is necessary for ApiCompat to run.  Ugh.  😄 Anyway, I've added those implementations with 28bd614ef8e3c7b3471a4d5775f201af146ba1ff.
